[receiver/mongodbreceiver] Add metric versioning checks to prevent partial errors

JonathanWamsley opened the pull request:
**Description:**
In use or in the integration tests, many partial errors are being emitted. A metric versioning and storage engine check can be applied before metrics are recorded to prevent these partial errors from occurring.
Figuring what partial errors were slipping through was a pain so a more detailed partial error message and logging was added to prevent a generic `could not find key for metric` message.

**Link to tracking Issue:**
#13155 

**Testing:**
- added test case for triggering partial error when metric paths are not valid
- added test case for triggering partial errors when a bad client response is received

**Documentation:**
The readme metrics section has mongodb version and storage engine requirements to get certain metrics.

Also while collecting `mongodb.global_lock.time`, it appears that mongodb [2.6](https://www.mongodb.com/docs/v2.6/reference/command/serverStatus/#serverStatus.globalLock.totalTime), [3.0](https://www.mongodb.com/docs/v3.0/reference/command/serverStatus/#serverStatus.globalLock.totalTime), [4.0](https://www.mongodb.com/docs/v4.0/reference/command/serverStatus/#serverstatus.globalLock.totalTime), and [5.0](https://www.mongodb.com/docs/v5.0/reference/command/serverStatus/#mongodb-serverstatus-serverstatus.globalLock.totalTime) all have the metric available so the old mongodb version check can be removed since it is available for all versions.

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/94d9600d623a0a0505e63bdfef8f9380942b4fbe/receiver/mongodbreceiver/metrics.go#L306-L309

JonathanWamsley committed:
add better error messages with tests for partial errors

JonathanWamsley committed:
add changelog

JonathanWamsley committed:
Merge branch 'main' into mongodbreceiver-add-metric-versioning

djaglowski commented on receiver/mongodbreceiver/README.md:
```diff
@@ -25,7 +25,6 @@ This receiver supports MongoDB versions:
 
 Mongodb recommends to set up a least privilege user (LPU) with a [`clusterMonitor` role](https://www.mongodb.com/docs/v5.0/reference/built-in-roles/#mongodb-authrole-clusterMonitor) in order to collect metrics. Please refer to [lpu.sh](./testdata/integration/scripts/lpu.sh) for an example of how to configure these permissions.
 
-Collecting metrics `mongodb.global_lock.time` and `mongodb.index.access.count` are only available for mongodb 4.0+.
```
I see that `mongodb.global_lock.time` was not moved to the list below. Is that because it actually is available regardless of version?

JonathanWamsley replied:
Yes, `mongodb.global_lock.time` is available on all supported versions 2.6, 3.0+, 4.0+ and 5.0 as shown in the pr description

djaglowski commented on receiver/mongodbreceiver/testdata/integration/expected.3_0.json:
```diff
@@ -43,23 +43,39 @@
                      },
                      "unit": "{databases}"
                   },
+                  {
+                     "description": "The time the global lock has been held.",
+                     "name": "mongodb.global_lock.time",
+                     "sum": {
+                        "aggregationTemporality": "AGGREGATION_TEMPORALITY_CUMULATIVE",
+                        "dataPoints": [
+                           {
+                              "asInt": "62216",
+                              "startTimeUnixNano": "1660317989210380000",
+                              "timeUnixNano": "1660318049220463000"
+                           }
+                        ],
+                        "isMonotonic": true
+                     },
+                     "unit": "ms"
+                  },
```
I'm guessing this corresponds to removal of this metric from the readme's version-specific notes.

JonathanWamsley replied:
Before when collecting global lock time, it was checking for version 4.0+ in the metric path `"globalLock", "totalTime"`. Else it would use the metric path `locks", ".", "timeAcquiringMicros"` but that path does not exist in 3.0 which was causing partial errors. This metric can be collected for 3.0 as shown in the pr description which is why `expected.3_0.json` got updated.

djaglowski reviewed (commented):
This is a lot to parse through. Is there a way to break this up into smaller PRs?

JonathanWamsley commented:
Hey @djaglowski  ,
I first added a partial error check shown in the [issue](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/13155) to panic before emitting metrics to see the partial errors when running the integration tests. That was not very helpful as they were `could not find key for metric;` I added different logging comments to quickly figure out which metrics were causing the partial errors and fixed them. This may be where 2 prs can be made like at [this stage](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/13109/commits/71b06e0d79fd51c897b0915826867793d892e8eb)

Ideally though I am not sure if that is the right way to do code development. I think adding in better logging and partial error messages  w/ tests is what I should have done originally which is why I implemented that. That specific metric messaging allowed me to make meaningful partial error tests for the client metric response and for each metric/attribute.

Having a way to collect partial errors in the integration tests would also be great but the consumer package MetricsSink struct populates the metrics and which I don't think there is a way to get partial errors from there.

If you want, I can split the pr into pr1 with the metric version updates like in the above link. Then pr2 with warning/partial error messages and unit tests. thoughts?

djaglowski commented:
@JonathanWamsley, that sounds reasonable, as long as we end up in the same place. It'll help me be able to follow what you're doing though.
